EasyLM/models/llama/llama_train_online_dpo.py

```diff
@@ -0,0 +1,544 @@
+    reward_model = FlaxLLaMAForSequenceClassification(llama_config_reward, dtype=get_float_dtype_by_name(FLAGS.dtype), _do_init=False)
+
+    print("Building optimizer...")
+    FLAGS.optimizer.adamw_optimizer.init_lr = 0.0
+    FLAGS.optimizer.adamw_optimizer.lr = FLAGS.lr
+    FLAGS.optimizer.adamw_optimizer.end_lr = FLAGS.lr
+    FLAGS.optimizer.adamw_optimizer.lr_warmup_steps = lr_warmup_grad_updates
+    policy_optimizer, policy_optimizer_info = OptimizerFactory.get_optimizer(FLAGS.optimizer)
+
+    print("Initializing training state and pjitting...")
+    def init_fn_policy(rng):
+        rng_generator = JaxRNG(rng)
+        params = policy_model.module.init(
+            input_ids=jnp.zeros((4, seq_length), dtype=jnp.int32),
+            position_ids=jnp.zeros((4, seq_length), dtype=jnp.int32),
+            attention_mask=jnp.ones((4, seq_length), dtype=jnp.int32),
+            rngs=rng_generator(llama_config_policy.rng_keys()),
+        )
+        return TrainState.create(params=params, tx=policy_optimizer, apply_fn=None)
+    def create_trainstate_from_params_policy(params):
+        return TrainState.create(params=params, tx=policy_optimizer, apply_fn=None)
+    train_state_shapes_policy = jax.eval_shape(init_fn_policy, next_rng()) # .params = {'params': {'transformer', 'lm_head'}} => .params = {'transformer', 'lm_head'}
+    train_state_partition_policy = match_partition_rules(LLaMAConfig.get_partition_rules(), train_state_shapes_policy)
+    shard_fns_policy, gather_fns_policy = make_shard_and_gather_fns(train_state_partition_policy, train_state_shapes_policy)
+    sharded_init_fn_policy = pjit(
+        init_fn_policy,
+        in_shardings=PS(),
+        out_shardings=train_state_partition_policy,
+    )
+    sharded_create_trainstate_from_params_policy = pjit(
+        create_trainstate_from_params_policy,
+        in_shardings=(train_state_partition_policy.params, ),
+        out_shardings=train_state_partition_policy,
+        donate_argnums=(0, ),
+    )
+
+    def init_fn_reward(rng):
+        rng_generator = JaxRNG(rng)
+        params = reward_model.module.init(
+            input_ids=jnp.zeros((4, seq_length), dtype=jnp.int32),
+            position_ids=jnp.zeros((4, seq_length), dtype=jnp.int32),
+            attention_mask=jnp.ones((4, seq_length), dtype=jnp.int32),
+            rngs=rng_generator(llama_config_reward.rng_keys()),
+        )
+        # I construct a train state, but later we will just use the params.
+        return TrainState.create(params=params, tx=policy_optimizer, apply_fn=None)
+    train_state_shapes_reward = jax.eval_shape(init_fn_reward, next_rng()) # .params = {'params': {'transformer', 'lm_head'}} => .params = {'transformer', 'lm_head'}
+    # TODO: do the partition rules correctly apply when you only return the params?
+    # something might break...
+    train_state_partition_reward = match_partition_rules(LLaMAConfig.get_partition_rules(), train_state_shapes_reward)
```

Comment:
I think `train_state_partition_reward.params` should work as the partition rule for only the params, so you're doing it correct. The trainstate creation in `init_fn_reward()` is a dummy thing, might be okay to get rid of but unnecessary ...

Reply:
Yeah. I tried to be smarter about this initially but it didn't quite work out because of something in the loading function... can't recall now. But ideally in the future not having to init a train state would be nice...